### PR TITLE
boot/fdehelper: add initrd fdehelper support

### PR DESCRIPTION
**Pull request by mvo5**
This commit adds minimal fdehelper support with support for revealing
a sealed key in initrd using systemd based basic confinement. This
helper will be used by snap-bootstrap when it unlocks the system on boot.

Input is passed via json on stdin
and the revealed key data is read via stdout. The actual storing
of the sealed keys will be done by snapd and secboot. The hooks
only gets key input and is expected to generate key output.

In future versions there will be complex boot-chain data passed
to the hooks but this initial version is intentionally kept simple.
It will still be useful for the simple ARM fde based use-cases.

For the initial-provision/updating of the key a "real" snap hook
from "meta/hooks/fde-setup" will be used (this is a separate PR).

This work is based on the spike work from Claudio (thanks you!)
from
https://github.com/snapcore/snapd/compare/master...cmatsuoka:spike/fde-helper-tpm#diff-9d89d75d6aa1bcb1db124c226af0d8e0R43

### CODE REVIEW

File: boot/fdehook/fdehook.go
Comment (zyga) [inline boot/fdehook/fdehook.go:57]: Nice!
I would love to do a TODO to investigate what else could be done in conjunction with the security team. Right now this is not a problem but, given the unique nature of this hook, we should be most paranoid about tightening security as much as we can get away with, without breaking the capability of the hook to operate.

File: boot/fdehook/fdehook.go
Comment (zyga) [inline boot/fdehook/fdehook.go:64]: That's a good start, if it's too restrictive we can simply iterate and open this up.

File: boot/fdehook/fdehook.go
Comment (zyga) [inline boot/fdehook/fdehook.go:50]: I was thinking if this should be done over the DBus API but I think this is sufficient and the extra preparation by `systemd-run` is non-trivial enough to keep this as a subprocess call. I would reverse this only if we need to monitor the precise exit status of the original process.

File: boot/fdehook/fdehook.go
Comment (zyga) [inline boot/fdehook/fdehook.go:66]: I would think about adding a timeout, so that the hook must unlock the partition in a reasonable amount of time or have the boot fail with a reasonable message.

File: boot/fdehook/fdehook.go
Comment (zyga) [inline boot/fdehook/fdehook.go:91]: Should those be prefixed with `SNAPD_` or something similar?

Comment (zyga) [inline boot/fdehook/fdehook.go:91]: Security paranoid me would prefer not to use the environment and instead pass the data in some other way but given that this runs in initrd I'm reluctant to say that this is probably okay.

We should ensure that the process, and all of its descendants, are dead though, to ensure that any information cannot leak by negligence.

File: boot/fdehook/fdehook.go
Comment (zyga) [inline boot/fdehook/fdehook.go:115]: Naming nitpick, we seem to use `command <subcommand>` for things like that.

**Review by zyga — COMMENTED** [review]
Quick first pass

File: boot/fdehook/fdehook.go
Comment (zyga) [inline boot/fdehook/fdehook.go:70]: This is so nice!

File: boot/fdehook/fdehook.go
Comment (cmatsuoka) [reply boot/fdehook/fdehook.go:80]: Note that we already have a case where access to a device will be needed.

File: boot/fdehook/fdehook.go
Comment (cmatsuoka) [reply boot/fdehook/fdehook.go:107]: It's passing the sealed (encrypted) key which is not a secret. The helper should process (decrypt) it to a form that can be used to unlock the data partition (or alternatively it wouldn't need to be passed and the helper could load the key from its location in the filesystem or some kind of internal storage that only the helper can access).

File: boot/fdehook/fdehook.go
Comment (pedronis) [inline boot/fdehook/fdehook.go:34]: what's the plan here, to have  two copies of the hook inside the kernel snap, one out the efi under meta? and one inside the initrd but at with some of the same parent tree?

Comment (mvo5) [reply boot/fdehook/fdehook.go:34]: That's a really good question. I think we will probably have to live with two copies. Alternatively we could split the helper into a "lock" and "unlock" helper. The later option might be clearer and it will also make very clear what part should be in the initrd and what part in the hooks dir. Wdyt?

Comment (pedronis) [reply boot/fdehook/fdehook.go:34]: also to be clear I don't think it should go to hooks if it's not called like the other hooks

File: boot/fdehook/fdehook.go
Comment (tsunghanliu) [inline boot/fdehook/fdehook.go:106]: I tried this branch and ran the run-fdehook test, then I found `systemd-run` didn't pass these environment variables to the helper.
So I think we need to pass the env variables to `fdeHookCmd` and add additional arguments such as `--setenv=FDE_SEALED_KEY=...` there instead.

File: boot/fdehook/fdehook.go
Comment (tsunghanliu) [inline boot/fdehook/fdehook.go:134]: Same issue of env variables as `Unlock()`

**Review by tsunghanliu — CHANGES_REQUESTED** [review]
Found issue of passing environment variables. Please check the comments

File: boot/fdehook/fdehook.go
Comment (mvo5) [reply boot/fdehook/fdehook.go:60]: Added a comment there.

File: boot/fdehook/fdehook.go
Comment (mvo5) [reply boot/fdehook/fdehook.go:131]: This got changed in the latest design and it's just "fde-reveal-key" now without further subcommands (for now)

File: boot/fdehook/fdehook.go
Comment (mvo5) [reply boot/fdehook/fdehook.go:49]: I moved this to 1m now

File: boot/fdehook/fdehook.go
Comment (mvo5) [reply boot/fdehook/fdehook.go:87]: Yeah, that is a good point, we need to account for this. I add at least a TODO

File: boot/fdehook/fdehook.go
Comment (mvo5) [reply boot/fdehook/fdehook.go:34]: I renamed fdehook -> fdehelper as this is not a hook in the snapd sense.

File: boot/fdehook/fdehook.go
Comment (mvo5) [reply boot/fdehook/fdehook.go:106]: Thanks for spotting this! I'm fixing this now and will improve the test

File: boot/fdehook/fdehook.go
Comment (mvo5) [reply boot/fdehook/fdehook.go:73]: I will chat with the security team about this. I'm not sure we need to really go all the way though because this is really about abuse/errors not about real security. The hook is part of the kernel and initrd so if someone wants to really break the security of the system there are many more and easier ways than to use the hook, i.e. just adding a a unit to the kernel intird or similar.

File: boot/fdehook/fdehook.go
Comment (mvo5) [reply boot/fdehook/fdehook.go:82]: Timeout added

File: boot/fdehook/fdehook.go
Comment (mvo5) [reply boot/fdehook/fdehook.go:107]: Not sure, I will leave this to the architects to decide :)

Comment (mvo5) [reply boot/fdehook/fdehook.go:107]: I'm not even sure we want this - OTOH it's nicer for the simple cases and when languages without good json support (like C) are used.

File: boot/fdehelper/fdehelper.go
Comment (mvo5) [inline boot/fdehelper/fdehelper.go:110]: XXX: use "fde-" prefix here?

**Comment by mvo5** [comment]
Closing as discussed with Samuele, the systemd-run part will be reused but the other bits are no longer up-to-date.

**Pull request closed without merging**
